### PR TITLE
Remove char[] allocations from RegistryKey and List copy to Array

### PR DESCRIPTION
**Pull request by Poppyto**
Replace char[] allocation with stackalloc memory (no need to rent for 256 chars)
Replace List<string> by string[] to avoid List.ToArray() copy (it must handle rare cases where there are more or less elements during the loop).
Avoid redondant verification EnsureNotDisposed() already called by SubKeyCount & ValueCount.

### CODE REVIEW

File: src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.Windows.cs [hunk 1]
```diff
@@ -387,45 +387,52 @@ private int InternalSubKeyCountCore()
...
                     }
                 }
-            }
-            finally
-            {
-                ArrayPool<char>.Shared.Return(name);
+
+                //less items than initially expected
+                if (cpt < names.Length)
+                    Array.Resize(ref names, cpt);
```
Comment (stephentoub) [inline src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.Windows.cs:432]:
```suggestion
                // Shrink array to fit found items, if necessary
                if (cpt < names.Length)
                {
                    Array.Resize(ref names, cpt);
                }
```

File: src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.Windows.cs [hunk 2]
```diff
@@ -387,45 +387,52 @@ private int InternalSubKeyCountCore()
...
+                null)) != Interop.Errors.ERROR_NO_MORE_ITEMS)
                 {
                     switch (result)
                     {
                         case Interop.Errors.ERROR_SUCCESS:
-                            names.Add(new string(name, 0, nameLength));
-                            nameLength = name.Length;
+
+                            if (cpt >= names.Length) // possible new item during loop
+                                Array.Resize(ref names, names.Length + 1);
```
Comment (stephentoub) [inline src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.Windows.cs:417]:
```suggestion
                            if (cpt >= names.Length) // possible new item during loop
                            {
                                Array.Resize(ref names, names.Length + 1);
                            }
```

**Review by stephentoub — COMMENTED** [review]
Thanks.  Can you share a perf test with numbers?

File: src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.Windows.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Poppyto) [reply src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.Windows.cs]: We could also read again SubKeyCount ? Doubling size for an potentially already 10k items array seems not an elegant solution.

[the remark this answers is not in the document]
Comment (Poppyto) [reply src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.Windows.cs]: I think this code would handle elegantly both situations : 

```
if (cpt >= names.Length) // possible new item during loop
{
	int newSize = InternalSubKeyCountCore();
	
	if(newSize > names.Length)
		Array.Resize(ref names, newSize);
	else // concurrent thread delete items, but we have to honor this last item
		Array.Resize(ref names, names.Length + 1);
}
```

[the remark this answers is not in the document]
Comment (stephentoub) [reply src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.Windows.cs]: That risks making this an infinite loop. I don't believe doubling in this case would be "inelegant"; it's a standard way of handling such things with a nice balance for all factors involved. But if you want to requery and only increase (not decrease) as part of the loop, we could, though I expect doing so will actually be more expensive than doubling in general; we don't expect the typical case to involve all that many subkeys, and syscalls are generally expensive.

**Comment by teo-tsirpanis** [comment]
Partially overlaps with #77996.

File: src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.Windows.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Poppyto) [reply src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.Windows.cs]: You are right, I send a fix.

**Review by stephentoub — APPROVED** [review]
Thanks.

**Pull request closed without merging**